dsls/sql/src/main/java/org/apache/beam/dsls/sql/transform/BeamBuiltinAggregations.java

```diff
@@ -33,275 +41,386 @@
    */
-  public static class Max<T extends Comparable<T>> extends BeamSqlUdaf<T, T> {
-    private BeamSqlRecordType accType;
+  public static final class Max<T extends Comparable<T>> extends BeamSqlUdaf<T, T, T> {
+    public static Max create(SqlTypeName fieldType) {
+      switch (fieldType) {
+        case INTEGER:
+          return new BeamBuiltinAggregations.Max<Integer>(fieldType);
+        case SMALLINT:
+          return new BeamBuiltinAggregations.Max<Short>(fieldType);
+        case TINYINT:
+          return new BeamBuiltinAggregations.Max<Byte>(fieldType);
+        case BIGINT:
+          return new BeamBuiltinAggregations.Max<Long>(fieldType);
+        case FLOAT:
+          return new BeamBuiltinAggregations.Max<Float>(fieldType);
+        case DOUBLE:
+          return new BeamBuiltinAggregations.Max<Double>(fieldType);
+        case TIMESTAMP:
+          return new BeamBuiltinAggregations.Max<Date>(fieldType);
+        case DECIMAL:
+          return new BeamBuiltinAggregations.Max<BigDecimal>(fieldType);
+        default:
+          throw new UnsupportedOperationException(
+              String.format("[%s] is not support in MAX", fieldType));
+      }
+    }
 
-    public Max(int outputFieldType) {
-      this.accType = BeamSqlRecordType.create(Arrays.asList("__max"),
-          Arrays.asList(outputFieldType));
+    private SqlTypeName fieldType;
+    private Max(SqlTypeName fieldType) {
+      this.fieldType = fieldType;
     }
 
     @Override
-    public BeamSqlRow init() {
+    public T init() {
       return null;
     }
 
     @Override
-    public BeamSqlRow add(BeamSqlRow accumulator, T input) {
-      return new BeamSqlRow(accType,
-          Arrays
-              .<Object>asList((accumulator == null || ((Comparable<T>) accumulator.getFieldValue(0))
-                  .compareTo(input) < 0)
-                      ? input : accumulator.getFieldValue(0)));
+    public T add(T accumulator, T input) {
+      return (accumulator == null || accumulator.compareTo(input) < 0) ? input : accumulator;
     }
 
     @Override
-    public BeamSqlRow merge(Iterable<BeamSqlRow> accumulators) {
-      T mergedV = (T) accumulators.iterator().next().getFieldValue(0);
-      while (accumulators.iterator().hasNext()) {
-        T v = (T) accumulators.iterator().next().getFieldValue(0);
+    public T merge(Iterable<T> accumulators) {
+      Iterator<T> ite = accumulators.iterator();
+      T mergedV = ite.next();
+      while (ite.hasNext()) {
+        T v = ite.next();
         mergedV = mergedV.compareTo(v) > 0 ? mergedV : v;
       }
-      return new BeamSqlRow(accType, Arrays.<Object>asList(mergedV));
+      return mergedV;
+    }
+
+    @Override
+    public T result(T accumulator) {
+      return accumulator;
     }
 
     @Override
-    public T result(BeamSqlRow accumulator) {
-      return (T) accumulator.getFieldValue(0);
+    public Coder<T> getAccumulatorCoder(CoderRegistry registry) throws CannotProvideCoderException {
```

Comment:
I was just thinking a static method in BeamBuiltinAggregations. It could take a registry parameter and a String describing the command name for the UnsupportedOperationException. What do you think?